### ISSUE TITLE
Add infer keyword

### ISSUE THREAD

**Issue by AlexanderFarkas** [dart-lang/language#2173]
Let:
```
class Entity<T> {
  T get id;
}

class EntityStore<T extends Entity<K>, K> {}

class Customer extends Entity<int> {
  final int id;
  Customer(this.id);
}
```

Calling Code:
`final store = EntityStore<Customer, int>();`

But type parameter K could already be inferred from T:
`final store = EntityStore<Customer>();`

I propose `infer` keyword for type parameters that could be inferred from other type parameters:
`class EntityStore<T extends Entity<K>, infer K> {}`

**Comment by lrhn** [comment]
Should type bounds really be type *patterns*?
```dart
class EntityStore<T extends Entity<var K>> { ... }
```
@munificient

**Comment by munificent** [comment]
That's a really interesting idea. What about in parameter lists?

```dart
foo(List<final T> someList) { ... }
```

Deciding when to use a generic method and when to use a type pattern seems like a tricky and subtle design choice.

**Comment by leafpetersen** [comment]
> What about in parameter lists?

My initial take would be to avoid this, and instead add the first class version of `extractTypeArguments` that we've discussed.  I think the latter provides a strict super set of this functionality?

**Comment by lrhn** [comment]
The distinction between `foo<T>(List<T> someList) { ... }` and `foo(List<final T> someList) { ... }` would be that the latter destructures the actual runtime type of the argument and extracts type arguments. The former takes a separate type as argument, which the list's element type must be a subtype of.

The pattern-in-parameter is no different from doing `foo(List<dynamic> someList) { if (someList is List<final T>) { ... }}`, and can possible even be desugared to that.

There is no way to separate the `List<final T>` type from the actual runtime type, which is also why there is no type *parameter* for the function.

The next question would be whether you can do `add(List<final T> list, T value) { list.add(value); }`. You probably *cannot*. That's dependent typing and you can't statically ensure its safety, so it can't be allowed in a *binding* pattern which relies on static type soundness. (Same likely applies to doing type matching inside record matching, the type can't be used again in the same match if it's a binding match.) It's probably possible to use it in a matching pattern, because then it can fail to match.

We probably can allow the return type to depend on the pattern-matched type variable, like `T mid(List<final T> list) => list[list.length ~/ 2];`, because the return statements happen inside the scope of the binding.

It might very well be that some generic methods would be better written as pattern-matching type-extractors.
I can't remember one right now, though.  (A lot of generic methods are only generic in order to allow an existential type, like `void apply<A>(void Function(A) f, A arg) => f(arg);`. Those still need the type parameter, because the type affects both parameters.)


So, a pattern like `List<final T extends num>` would both constrain the type that it can match *and* extract the runtime element type if it matches.

**Comment by munificent** [comment]
> > What about in parameter lists?
> 
> My initial take would be to avoid this, and instead add the first class version of `extractTypeArguments` that we've discussed. I think the latter provides a strict super set of this functionality?

Yeah, the patterns proposal already covers this use case. But thinking about using type patterns in bounds made me wonder about using them in other places where type parameters can occur when binding is happening and parameter lists are another example of that. (This, of course, also raises the question of allowing other kinds of patterns in parameter lists.)

> The pattern-in-parameter is no different from doing `foo(List<dynamic> someList) { if (someList is List<final T>) { ... }}`, and can possible even be desugared to that.

Yeah, I think this is the right approach. Allowing type patterns directly in the parameter list seems like a can of worms.

**Comment by lrhn** [comment]
>> The pattern-in-parameter is no different from doing `foo(List<dynamic> someList) { if (someList is List<final T>) { ... }}`, and can possible even be desugared to that.

> Yeah, I think this is the right approach. Allowing type patterns directly in the parameter list seems like a can of worms.

I was actually using that as an argument for allowing patterns in parameter lists. Pattern-parameters are statically typed irrefutable binding constructs, just like what variable declarations become with patterns, so I don't think there should be any problem, and forcing people to declare the un-patterned variable and then immediately pattern-match it inside the function, is just adding overhead.

**Comment by pedromassango** [comment]
I think we should close this in favor of https://github.com/dart-lang/language/issues/620 as they address the exact same issue but with different proposed solutions. Or the other way around.

**Comment by pedromassango** [comment]
btw, is this ticket (or #620) something that is being considered/talked about by the team? 👀

**Comment by TimWhiting** [comment]
Cross referencing with the issue related to patterns https://github.com/dart-lang/language/issues/2221

**Comment by shtse8** [comment]
In my use case, I want to wrap a list in generic type with a class without knowing what inner type is.

```dart
T wrap(T obj) {   
  if (obj is List) obj = ReactiveList(obj) as T
  return obj;
}
```
where `ReactiveList<T> extends List<T>`.

Resulting in:
```
_CastError (type 'ReactiveList<dynamic>' is not a subtype of type 'List<int>' in type cast.
```

**Comment by AlexanderFarkas** [comment]
@shtse8 Why not just
```dart
ReactiveList<T> wrap<T>(Iterable<T> list) {
  return ReactiveList(lsit);
}
```

**Comment by shtse8** [comment]
> @shtse8 Why not just
> 
> ```dart
> ReactiveList<T> wrap<T>(Iterable<T> list) {
>   return ReactiveList(list);
> }
> ```

because T is probably other types, not only list.
`wrap` is one of the methods in class. the value stored in the class is generic type but I just need to wrap it if the type is `List`.
